File: app/src/main/kotlin/com/absinthe/libchecker/utils/manifest/HiddenPermissionsReader.java
package com.absinthe.libchecker.utils.manifest;

import androidx.collection.ArrayMap;


import java.io.File;
import java.io.IOException;
import java.util.Map;

import pxb.android.Res_value;
import pxb.android.axml.NodeVisitor;

public class HiddenPermissionsReader {
  private final ArrayMap<String, Object> permissionMap = new ArrayMap<>();

  private HiddenPermissionsReader(File apk) {
    ManifestReader.acceptManifest(apk, () -> new ManifestTagVisitor(null));
  }

  public static Map<String, Object> getHiddenPermissions(File apk) throws IOException {
    return new HiddenPermissionsReader(apk).permissionMap;
  }

  private class ManifestTagVisitor extends NodeVisitor {
    public ManifestTagVisitor(NodeVisitor child) {
      super(child);
    }

    @Override
    public NodeVisitor child(String ns, String name) {
      NodeVisitor child = super.child(ns, name);
      if ("uses-permission".equals(name)) {
        return new PermissionVisitor(child);
      }
      return child;
    }
  }

  private class PermissionVisitor extends NodeVisitor {
    public String name = null;
    public Object maxSdkVersion = null;

    public PermissionVisitor(NodeVisitor child) {
      super(child);
    }

    @Override
    public void attr(String ns, String name, int resourceId, String raw, Res_value value) {
      if ("name".equals(name) && value.type == Res_value.TYPE_STRING) {
        this.name = value.toString();
      } else if ("maxSdkVersion".equals(name) && value.type == Res_value.TYPE_INT_DEC) {
        this.maxSdkVersion = value.data;
      }
      super.attr(ns, name, resourceId, raw, value);
    }

    @Override
    public void end() {
      if (name != null && maxSdkVersion != null) {
        permissionMap.put(name, maxSdkVersion);
      }
      super.end();
    }
  }
}

File: app/src/main/kotlin/com/absinthe/libchecker/utils/manifest/HiddenPermissionsReader.kt
package com.absinthe.libchecker.utils.manifest

import androidx.collection.ArrayMap
import com.absinthe.libchecker.compat.IZipFile
import com.absinthe.libchecker.compat.ZipFileCompat
import pxb.android.Res_value
import pxb.android.axml.AxmlReader
import pxb.android.axml.AxmlVisitor
import pxb.android.axml.NodeVisitor
import timber.log.Timber
import java.io.ByteArrayOutputStream
import java.io.File
import java.io.IOException
import java.io.InputStream

class HiddenPermissionsReader private constructor(apk: File) {
    private val permissionMap = ArrayMap<String, Any>()

    init {
        try {
            ZipFileCompat(apk).use { zip ->
                val inputStream = zip.getInputStream(zip.getEntry("AndroidManifest.xml"))
                val bytes = getBytesFromInputStream(inputStream)
                val reader = AxmlReader(bytes ?: ByteArray(0))
                reader.accept(object : AxmlVisitor() {
                    override fun child(ns: String?, name: String?): NodeVisitor {
                        val child = super.child(ns, name)
                        return ManifestTagVisitor(child)
                    }
                })
            }
        } catch (e: Exception) {
            Timber.e(e)
        }
    }

    companion object {
        @JvmStatic
        @Throws(IOException::class)
        fun getHiddenPermissions(apk: File): Map<String, Any> {
            return HiddenPermissionsReader(apk).permissionMap
        }

        private fun getBytesFromInputStream(inputStream: InputStream?): ByteArray? {
            if (inputStream == null) return null
            
            return try {
                ByteArrayOutputStream().use { bos ->
                    val b = ByteArray(1024)
                    var n: Int
                    while (inputStream.read(b).also { n = it } != -1) {
                        bos.write(b, 0, n)
                    }
                    bos.toByteArray()
                }
            } catch (e: Exception) {
                Timber.w(e)
                null
            }
        }
    }

    private inner class ManifestTagVisitor(child: NodeVisitor?) : NodeVisitor(child) {
        override fun child(ns: String?, name: String?): NodeVisitor {
            val child = super.child(ns, name)
            return if ("uses-permission" == name) {
                PermissionVisitor(child)
            } else child
        }
    }

    private inner class PermissionVisitor(child: NodeVisitor?) : NodeVisitor(child) {
        var name: String? = null
        var maxSdkVersion: Any? = null

        override fun attr(ns: String?, name: String?, resourceId: Int, raw: String?, value: Res_value?) {
            if ("name" == name && value?.type == Res_value.TYPE_STRING) {
                this.name = value.toString()
            } else if ("maxSdkVersion" == name && value?.type == Res_value.TYPE_INT_DEC) {
                this.maxSdkVersion = value.data
            }
            super.attr(ns, name, resourceId, raw, value)
        }

        override fun end() {
            val permName = name
            val maxSdk = maxSdkVersion
            if (permName != null && maxSdk != null) {
                permissionMap[permName] = maxSdk
            }
            super.end()
        }
    }
}